### PR TITLE
OpenZiti download components role

### PR DESCRIPTION
**Pull request by ares-b**
Hello,

Submitting a role to download openziti components.

This role can : 

- Download OpenZiti Components (ziti CLI, ziti-controller, ziti-router, ziti-edge & ziti-console), cache them on a cache remote host or on localhost
- Put these components on remote servers

We might need to discuss about where to store Ziti binaries and ZAC source code by default. I've put them in `/opt/openziti` by default (can be changed by variables tho), but I'm not sure that's the best location by default.

Also, I need your opinion on whether to use group names, group vars or host vars to specify on which host deploy which OpenZiti component. Atm, the role is using group names, but again, I'm not sure that's the best choice.

One last thing, I'm planning on handling `become_password`s with varibles, we need like 2 become_passwords : 
- One for Ansible cache server (could be a remote host or localhost)
- One for OpenZiti Network hosts

I'm thinking about asking users to give those as input in their playbooks using ansible-vault, AWX or a vault. What do you guys think ?

The PR also contains a github actions CI `.github/workflows/molecule-ci.yml`, it has two jobs, the first one goes over the repository and catches every directory having a sub-directory named `molecule`. The second, runs `molecule test` against every one of those directories.

Feel free to give me feedback, I'll take it into consideration and change what needs to be changed.

Regards,
Arslane BAHLEL

### CODE REVIEW

**Comment by ares-b** [comment]
Hi @qrkourier, I just pushed the changes you asked for.

As you may have noticed, I've done everything with Ansible existing modules, I think that here's a few tasks that can be simplified with a custom module such as the `Select binaries and remote_server_path from openziti components variable` on `push_to_remote.yaml` file.

I was wondering, if you guys would be okay with me making custom module to make the ansible role more succinct.

**Comment by qrkourier** [comment]
@ares-b I'm definitely in favor of custom modules for the reason you mentioned.

Here's what happened when I tried to run the demo playbook.

```bash
❯ ansible-playbook community.openziti.demo_openziti_download -u ec2-user -i 13.57.252.159,

 ____________________________________________________________
/ TASK [community.openziti.openziti_download : Make sure tar \
\ is installed on cache server]                              /
 ------------------------------------------------------------

An exception occurred during task execution. To see the full traceback, use -vvv. The error was: NoneType: None
fatal: [13.57.252.159 -> localhost]: FAILED! => {"changed": false, "msg": "Could not detect which package manager to use. Try gathering facts or setting the \"use\" option."}
```

**Comment by ares-b** [comment]
@qrkourier That's because `ansible.builtin.package` needs to be executed on sudo mode, and I'm getting that you've left `openziti_cache_localhost = true`. You need to launch you playbook with `--ask-become-pass` or `-K`, then you enter your machine's become password.

```
ansible-playbook community.openziti.demo_openziti_download -u ec2-user -i 13.57.252.159, -K
```

**Comment by qrkourier** [comment]
@ares-b This appears to be a stray string in the README:
> 5 Allee Saint Exupery 92390 Villeneuve La Garenne

**Comment by qrkourier** [comment]
@ares-b I tried to get this working with a remote cache server because, personally, I would strongly prefer to avoid using `become` on localhost. In this case there is no become password for ec2-user on Amazon Linux.

```yaml
# hosts.yml
all:
  children:
    dev_network:
      hosts:
        host_one:
          ansible_host: 13.57.252.159
          ansible_ssh_user: ec2-user
          ansible_become_pass: ''
          openziti_components:
            - ziti-controller
            - ziti-router
            - ziti-tunnel
            - ziti
            - ziti-console
            - ziti-edge-tunnel
        cache_server:
          ansible_host: 13.57.252.159
          ansible_ssh_user: ec2-user
          ansible_become_pass: ''
```

```bash
$ ansible-playbook community.openziti.demo_openziti_download -i ./hosts.yml
 ____________________________________________________________
/ TASK [community.openziti.openziti_download : Make sure tar \
\ is installed on cache server]                              /
 ------------------------------------------------------------
                                                                                             
fatal: [host_one -> localhost]: FAILED! => {"msg": "{{ my_localhost_sudo_pass_from_secrets }}: 'my_localhost_sudo_pass_from_secrets' is undefined"}
...

**Comment by ares-b** [comment]
Hi @qrkourier 

> This appears to be a stray string in the README

Right, joys of copy/pasting haha

> I tried to get this working with a remote cache server because, personally, I would strongly prefer to avoid using become on localhost. In this case there is no become password for ec2-user on Amazon Linux. 

Can you try and remove `localhost_become_pass` from your playbook ? I've left it by default to `"{{ my_localhost_sudo_pass_from_secrets }}"` and the error says that it can't locate variable `my_localhost_sudo_pass_from_secrets`.

Try running it with this playbook : 
```
- hosts: '{{ target | default("all") }}'

  tasks:
    - ansible.builtin.include_role:
        name: openziti_download
      vars:
        openziti_cache_localhost: false

```

**Comment by ares-b** [comment]
> I found the correct components installed with remote cache mode on RHEL8, Fedora34, Ubuntu20, and Amazon Linux.

Alright, it should also work with Debian, CentOS and Oracle (and probably every other Unix distribution).

If you want to test it with localhost cache, make sure you have tar, zip and git installed and disable these tasks : 

- Make sure tar is installed on cache server
- Make sure zip is installed on cache server
- Make sure git is installed on cache server

**Comment by qrkourier** [comment]
@ares-b Local cache testing succeeded with Debian11, SUSE15, and Oracle8. I tagged and did `--skip-tags` for those three tasks that check required packages on the cache server.

**Comment by qrkourier** [comment]
@ares-b I found that `pip install "molecule[lint]"` will install all three PyPi packages.

**Comment by ares-b** [comment]
@qrkourier The error you're getting is weird, can you try this :
```
cd community.openziti/roles/openziti_download
molecule test
```
molecule test runs the default scenario, if you want to try all scenarios, make it `molecule test --all`

File: community.openziti/roles/openziti_download/README.md [hunk 1]
```diff
@@ -19,6 +19,9 @@ Either specify your localhost sudo password with `-K` or `--ask-become-pass` opt
 
 When using a remote host as the `cache-server` (i.e `openziti_cache_localhost` = `false`), you should provide the sudo password in host variables.
 
+### Avoid sudo on cache_server/localhost
+If you want to avoid prividing become password for your cache_server or localhost, just make sur you've tar, zip and git packages installed and skip the tag `cache_server_packages` by running the playbook with `--skip-tags cache_server_packages`.
```
Comment (qrkourier) [inline community.openziti/roles/openziti_download/README.md:23]: typo
> just make sure

Comment (ares-b) [reply community.openziti/roles/openziti_download/README.md]: fixed

**Comment by qrkourier** [comment]
@ares-b I'm able to run the default tests when I simply say `molecule test` in the role dir, as you suggested.

The next issue I found is that this test appears to get stuck. Perhaps it is waiting for a become password?

```bash
❯ docker ps
CONTAINER ID   IMAGE                                          COMMAND                  CREATED         STATUS         PORTS     NAMES
28e8ef9a1579   molecule_local/docker.io/amd64/oraclelinux:8   "bash -c 'while true…"   3 minutes ago   Up 3 minutes             oracle_amd64
44fc7d577800   molecule_local/docker.io/amd64/ubuntu          "bash -c 'while true…"   3 minutes ago   Up 3 minutes             ubuntu_amd64

❯ molecule test
# ...
TASK [openziti_download : Make sure tar is installed on cache server] **********                         

^C                                                                                                                                                                                                                
Aborted!                                                                                                 
```

**Comment by ares-b** [comment]
> @ares-b I'm able to run the default tests when I simply say `molecule test` in the role dir, as you suggested.
> 
> The next issue I found is that this test appears to get stuck. Perhaps it is waiting for a become password?
> 
> ```shell
> ❯ docker ps
> CONTAINER ID   IMAGE                                          COMMAND                  CREATED         STATUS         PORTS     NAMES
> 28e8ef9a1579   molecule_local/docker.io/amd64/oraclelinux:8   "bash -c 'while true…"   3 minutes ago   Up 3 minutes             oracle_amd64
> 44fc7d577800   molecule_local/docker.io/amd64/ubuntu          "bash -c 'while true…"   3 minutes ago   Up 3 minutes             ubuntu_amd64
> 
> ❯ molecule test
> # ...
> TASK [openziti_download : Make sure tar is installed on cache server] **********                         
> 
> ^C                                                                                                                                                                                                                
> Aborted!                                                                                                 
> ```

Yeah, basically the default scenario launches the role with your localhost as cache server. You can launch with `molecule test --skip-tags=cache_server_packages`

**Comment by qrkourier** [comment]
@ares-b I was able to run the default tests like this, but the same test gets stuck. It's as if the skip-tags option was ignored.

```bash
# note the added POSIX -- to stop parsing positionals as options for `test` subcommand, and instead pass through to `ansible-playbook`
molecule test -- --skip-tags=cache_server_packages
```

When running the remote_cache test suite I get that same unfamiliar error.

```bash
$ molecule test ./molecule/remote_cache/molecule.yml                                     
# ...
INFO     Running default > syntax
ERROR! A playbook must be a list of plays, got a <class 'ansible.parsing.yaml.objects.AnsibleMapping'> instead

The error appears to be in '/home/kbingham/Sites/netfoundry/github/openziti-ansible-collection/community.openziti/roles/openziti_download/molecule/remote_cache/molecule.yml': line 3, column 1, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


dependency:
^ here
```

**Comment by ares-b** [comment]
@qrkourier both of scenarios are working perfectly on my localhost and with github actions https://github.com/ares-b/openziti-ansible-collection/actions

Can you try this :
```
cd community.openziti/roles/openziti_download
molecule test -- --skip-tags=cache_server_packages
```
and 
```
cd community.openziti/roles/openziti_download
molecule test -s remote_cache
```

I don't think that `molecule test ./molecule/remote_cache/molecule.yml` is supposed to work, you always need to be into the "working directory" before launching a `molecule <cmd>` command

**Review by qrkourier — APPROVED** [review]
Thanks for educating me about the syntax for selecting a Molecule scenario. Now the remote_cache scenario succeeds as expected.

I overcame the issue with `molecule test -- --skip-tags=cache_server_packages` by reseting my branch to match yours. Thanks!

**Pull request merged**